File: game_store.py
from typing import Optional, Dict, Any
from supabase import create_client, Client
import os
from dotenv import load_dotenv
from game_engine import MetaBoard
from minimax_agent import MinimaxAgent

# Load environment variables
load_dotenv('env')


class GameData:
    def __init__(self, board: MetaBoard, ai_agent: MinimaxAgent):
        self.board = board
        self.ai_agent = ai_agent


class GameStore:
    def __init__(self):
        supabase_url = os.getenv("SUPABASE_URL")
        supabase_key = os.getenv("SUPABASE_KEY")
        
        if not supabase_url or not supabase_key:
            raise ValueError("SUPABASE_URL and SUPABASE_KEY must be set in env file")
        
        print(f"Initializing game store with Supabase URL: {supabase_url[:30]}...")
        self.supabase: Client = create_client(supabase_url, supabase_key)
        self.games: Dict[str, GameData] = {}

    def create_new_game(self, user_id: str) -> str:
        """Create a new game in Supabase and return the game ID"""
        from uuid import uuid4
        
        game_id = str(uuid4())
        new_board = MetaBoard()
        
        # Get state and convert to dict
        state = new_board.get_state()
        game_state = {
            "boards": [{"board": b.board, "winner": b.winner, "available_moves": b.available_moves} for b in state.boards],
            "meta_board": state.meta_board,
            "game_winner": state.game_winner,
            "next_board": state.next_board,
            "available_boards": state.available_boards,
            "move_history": state.move_history,
        }
        
        # Insert into Supabase
        self.supabase.table("games").insert({
            "id": game_id,
            "user_id": user_id,
            "game_state": game_state,
        }).execute()
        
        # Store in memory
        self.games[game_id] = GameData(
            board=new_board,
            ai_agent=MinimaxAgent(2, 4)
        )
        
        return game_id

    def get_game(self, game_id: str) -> Optional[GameData]:
        """Get game from cache or Supabase"""
        # Check cache first
        if game_id in self.games:
            return self.games[game_id]
        
        # Load from Supabase
        result = self.supabase.table("games").select("*").eq("id", game_id).execute()
        
        if not result.data:
            return None
        
        data = result.data[0]
        state = data["game_state"]
        
        # Reconstruct the board
        board = MetaBoard()
        
        # Reconstruct boards
        for i, b_state in enumerate(state["boards"]):
            board.boards[i].board = b_state["board"]
            board.boards[i].winner = b_state["winner"]
        
        board.meta_board = state["meta_board"]
        board.game_winner = state["game_winner"]
        board.next_board = state["next_board"]
        board.move_history = state["move_history"]
        
        game_data = GameData(
            board=board,
            ai_agent=MinimaxAgent(2, 4)
        )
        
        # Cache it
        self.games[game_id] = game_data
        
        return game_data

    def update_game(self, game_id: str, game_data: GameData) -> None:
        """Update game in Supabase"""
        self.games[game_id] = game_data
        
        state = game_data.board.get_state()
        
        # Convert dataclass to dict for JSON serialization
        game_state = {
            "boards": [{"board": b.board, "winner": b.winner, "available_moves": b.available_moves} for b in state.boards],
            "meta_board": state.meta_board,
            "game_winner": state.game_winner,
            "next_board": state.next_board,
            "available_boards": state.available_boards,
            "move_history": state.move_history,
        }
        
        self.supabase.table("games").update({
            "game_state": game_state,
        }).eq("id", game_id).execute()

    def delete_game(self, game_id: str) -> None:
        """Delete game from Supabase and cache"""
        self.supabase.table("games").delete().eq("id", game_id).execute()
        if game_id in self.games:
            del self.games[game_id]


# Global instance
game_store = GameStore()

File: lib/game-store.ts
import { MetaBoard, MiniBoard } from "./game-engine"
import { MinimaxAgent } from "./minimax-agent"
import { supabase } from "./supabase"

interface GameData {
  board: MetaBoard
  ai_agent: MinimaxAgent
}

const games: Map<string, GameData> = new Map()

// Load game from Supabase
export async function loadGameFromDB(gameId: string): Promise<GameData | null> {
  const { data, error } = await supabase
    .from("games")
    .select("*")
    .eq("id", gameId)
    .single()

  if (error || !data) {
    return null
  }

  const board = new MetaBoard()
  const state = data.game_state as any

  // Reconstruct the boards array
  board.boards = state.boards.map((b: any) => {
    const miniBoard = new MiniBoard()
    miniBoard.board = [...b.board]
    miniBoard.winner = b.winner
    return miniBoard
  })
  board.meta_board = [...state.meta_board]
  board.game_winner = state.game_winner
  board.next_board = state.next_board
  board.move_history = [...state.move_history]

  return {
    board,
    ai_agent: new MinimaxAgent(2, 4),
  }
}

// Save game to Supabase
export async function saveGameToDB(gameId: string, gameData: GameData): Promise<void> {
  await supabase
    .from("games")
    .update({
      game_state: gameData.board.get_state(),
      updated_at: new Date().toISOString(),
    })
    .eq("id", gameId)
}

// Create new game in Supabase
export async function createNewGame(userId: string): Promise<string> {
  const newBoard = new MetaBoard()
  const { data, error } = await supabase
    .from("games")
    .insert({
      user_id: userId,
      game_state: newBoard.get_state(),
    })
    .select()
    .single()

  if (error || !data) {
    throw new Error("Failed to create game in database")
  }

  return data.id
}

// Get game - loads from cache or Supabase
export async function getGame(gameId: string): Promise<GameData | null> {
  if (games.has(gameId)) {
    return games.get(gameId)!
  }

  const gameData = await loadGameFromDB(gameId)
  if (gameData) {
    games.set(gameId, gameData)
  }
  return gameData
}

// Update game in memory and database
export async function updateGame(gameId: string, gameData: GameData): Promise<void> {
  games.set(gameId, gameData)
  await saveGameToDB(gameId, gameData)
}

// Delete game from database and cache
export async function deleteGame(gameId: string): Promise<void> {
  await supabase.from("games").delete().eq("id", gameId)
  games.delete(gameId)
}
